coriolis/providers/replicator.py

```diff
@@ -283,6 +283,114 @@ def init_replicator(self):
+        [{
+            'device-path': '/dev/xvdf',
+            'device-name': 'xvdf',
+            'size': 10737418240,
+            'checksum-algorithm': 'sha256',
+            'chunk-size': 10485760,
+            'logical-sector-size': 512,
+            'physical-sector-size': 512,
+            'alignment-offset': 0,
+            'has-mounted-partitions': False,
+            'checksum-status': {
+                'status': 'running',
+                'total-chunks': 1024,
+                'completed-chunks': 0,
+                'percentage': 0},
+            'partitions': [{
+                'name': 'xvdf1',
+                'sectors': 20969439,
+                'uuid': '73e9659d-2fd9-46ca-a341-5a8637c416ee',
+                'fs': 'ext4',
+                'start-sector': 2048,
+                'end-sector': 20971486,
+                'alignment-offset': 0}]
+        }]
+        """
+        return self._cli.get_status()
+
+    def attach_new_disk(
+            self, disk_id, attachf, previous_disks_status=None):
+        """ Returns the volumes_info for the disk attached by running
+        `attachf`. This is achieved by comparing the disk statuses before and
+        after the execution of the attachment operation.
+
+        param disk_id: str(): the 'disk_id' of the info from self._volumes_info
+            for the disk which shall be attached by `attachf()`.
+        param attachf: fun(): argument-less function which attaches the disk.
+            The function should perform any waits required to reasonably expect
+            the worker OS to have noticed the new disk, or reboot the worker VM
+            and re-run `init_replicator()` if deemed necessary.
+        param previous_disks_status: dict(): previous status of the disks as
+            returned by `get_current_disks_status()`.
+
+        return: dict(): returns the volumes_info associated to the new disk.
+        """
+        # check if volume with given ID is present in self._volumes_info:
+        matching_vols = [
+            vol for vol in self._volumes_info
+            if vol['disk_id'] == disk_id]
+        if not matching_vols:
+            raise exception.CoriolisException(
+                "No information regarding volume with ID '%s'. "
+                "Cannot track its attachment." % disk_id)
+        elif len(matching_vols) > 2:
+            raise exception.CoriolisException(
+                "Multiple volumes info with ID '%s' found: %s" % (
+                    disk_id, matching_vols))
+        vol_info = matching_vols[0]
+
+        # get/refresh current device paths:
+        if not previous_disks_status:
+            previous_disks_status = self._cli.get_status()
+        previous_device_paths = [
+            dev['device-path'] for dev in previous_disks_status]
+
+        # run attachment function and get new device paths:
+        attachf()
+        new_disks_status = self._cli.get_status()
```

Comment:
This should probably be polled several times. Devices may be detected slower than expected.

Reply:
As mentioned in the docstring, it's the responsibility of `attachf` to wait as much as it considers necessary.

Reply:
How would it know without checking with the replicator status if the disk has appeared on the server or not?